Sources/MongoSwift/MongoCollection+Write.swift

```diff
@@ -21,24 +21,8 @@ extension MongoCollection {
      */
     @discardableResult
     public func insertOne(_ value: CollectionType, options: InsertOneOptions? = nil) throws -> InsertOneResult? {
-        let document = try self.encoder.encode(value).withID()
-        let opts = try self.encoder.encode(options)
-        var error = bson_error_t()
-        let reply = Document()
-        guard mongoc_collection_insert_one(self._collection, document.data, opts?.data, reply.data, &error) else {
-            throw getErrorFromReply(bsonError: error, from: reply)
-        }
-
-        guard isAcknowledged(options?.writeConcern) else {
-            return nil
-        }
-
-        guard let insertedId = try document.getValue(for: "_id") else {
-            // we called `withID()`, so this should be present.
-            fatalError("Failed to get value for _id from document")
-        }
-
-        return InsertOneResult(insertedId: insertedId)
+        let operation = InsertOneOperation(collection: self, value: value, options: options)
```

Comment:
did you try to make a `BulkOperation` that serves as the basis for all write operations? It seems to me we could do something like:
```
let ops = [
  InsertOneOperation(collection: self, value: value, options: options)
]

let operation = BulkOperation(operations: ops)
return try operation.execute();
```

Each operation outside of `BulkOperation` would be like a pseudo-operation, and not have a functional `execute` (or even necessarily conform to `Operation`?). `BulkOperation` could then be responsible for switching on the well-defined operation types to determine which libmongoc method to call. 

Food for thought.

Reply:
what if we just updated all the write methods to do something like [`insertMany`](https://github.com/mongodb/mongo-swift-driver/blob/c78822f68c0522c701572d96753ed573f10b36e9/Sources/MongoSwift/MongoCollection%2BWrite.swift#L61)? construct a corresponding `BulkWriteModel`, convert the options, and call  `bulkWrite` with that. and then bulkWrite already handles all those different types of writes and we won't be calling all the operation specific methods anymore. (unless if we want to special case when only model is provided, but not sure it matters?) 

then possibly we would move the logic for each of these methods into corresponding operations types. 

we already have a `BulkWriteOperation` that `bulkWrite` constructs and executes, which will be made to conform to `Operation` in a future PR.

Reply:
Note that if we update the single write methods to use bulk writes under the hood, we'll need to do some work to parse any `.bulkWriteError`s back into regular `.writeError`s.

Reply:
good point. as the error expert... how much work do you think that is / do you foresee any issues there?

Reply:
There's a couple of ways to do it, the easiest being just to catch the bulk write error and translate it to a write error before throwing it again. They're easily convertible since a `.bulkWriteError` is just a `.writeError` that can contain multiple `WriteError`s.

Reply:
Hah yes, that sounds like _exactly_ what I was proposing. Thanks for pointing out the obvious here, let's definitely explore that path!